File: 0034-在排序数组中查找元素的第一个和最后一个位置/Solution.py
class Solution:
    def searchRange(self, nums, target):
        left = self.find_left_index(nums, target)
        right = -1 if left == -1 else self.find_right_index(nums, target)
        return [left, right]

    def find_left_index(self, nums, target):
        left, right = 0, len(nums)-1
        while(left <= right):
            mid = (left + right) // 2
            if nums[mid] == target and (mid == 0 or nums[mid-1] != target):
                return mid
            if nums[mid] >= target:
                right = mid - 1
            else:
                left = mid + 1
        return -1

    def find_right_index(self, nums, target):
        left, right = 0, len(nums)-1
        while(left <= right):
            mid = (left + right) // 2
            if nums[mid] == target and (mid == len(nums)-1 or nums[mid+1] != target):
                return mid
            if nums[mid] <= target:
                left = mid + 1
            else:
                right = mid - 1     
        return -1

File: 0034-在排序数组中查找元素的第一个和最后一个位置/Solution.java
class Solution {
    public int[] searchRange(int[] nums, int target) {
        int left = 0, right= nums.length - 1;
        int[] res = new int[]{-1, -1};       
        int mid;
        while(left <= right){
            mid = (left + right) / 2;
            if(nums[mid] == target){
                // find min index and max index
                int min = mid - 1, max = mid + 1;
                while(min >= 0 && nums[min] == target){ min--; };
                while(max < nums.length && nums[max] == target){ max++; };
                res[0] = min + 1;
                res[1] = max - 1;
                return res;

            }else if(nums[mid] < target){
                left = mid + 1;
            }else{
                right = mid - 1;
            }
        }
        return res;
    }
}